### ISSUE TITLE
error: no matching package named `...` found

### ISSUE THREAD

**Issue by srid** [kolloch/crate2nix#179]
I get this really strange error but only in flakes. `nix-build` is fine, but `nix run` breaks. And it only happens if the git working copy is not dirty.

```
++ crate2nix generate -f ./Cargo.toml -o Cargo-generated.nix -h /nix/store/0v78qgpw835pb2ri7r1l1v40linxhi5a-himalaya-crate2nix/crate-hashes.json
Error: while retrieving metadata about ./Cargo.toml: Error during execution of `cargo metadata`: error: no matching package named `clap` found
location searched: registry `https://github.com/rust-lang/crates.io-index`
required by package `himalaya v0.2.2 (/build/source)`

crate2nix failed.
```

The source tree to reproduce this is here: https://github.com/srid/himalaya/commit/fce129d6a36f8b535ce042ddcd611b4a73120a88

By the way, I came across crate2nix [via a conversation on reddit](https://www.reddit.com/r/rust/comments/mmbfnj/nixifying_a_rust_project/gtrmn73/?utm_source=reddit&utm_medium=web2x&context=3), which has further context on this problem.

**Comment by srid** [comment]
And I'm positive that this has nothing to do with the package in question, `clap`.

Confirmed here,

- https://github.com/srid/bouncy/runs/2293288367
- https://github.com/srid/bouncy/runs/2293424388

**Comment by srid** [comment]
@FintanH No, I no longer use Rust. Still using Haskell :-)

**Comment by Ericson2314** [comment]
@FintanH try using strace to get the underlying `cargo metadata` command? This might be an issue for upstream.

**Comment by FintanH** [comment]
> No, I no longer use Rust. Still using Haskell :-)

lol, very envious :)

> try using strace to get the underlying cargo metadata command? This might be an issue for upstream.

@Ericson2314 thanks for the suggestion! I gave that a go, but it seems like `crate2nix` is using the `cargo_metadata` package :sweat: I couldn't see anything out of the ordinary happening.

Two things that are weird:
1. If I drop into a shell `nix-shell -p crate2nix` and run `crate2nix generate` it works perfectly fine.
2. If I commit `Cargo.lock` it works perfectly fine.

So it seems like a combination of not having a `Cargo.lock` with this flake setup is messing things up. I'd rather not have a `Cargo.lock` in the repository because it's a library package.

**Comment by FintanH** [comment]
I did some digging by pointing to my own crate2nix fork and building off of a custom revision. I printed the command that gets executed:
```
cargo metadata --format-version 1 --manifest-path ./Cargo.toml --all-features --locked
```

The weird thing is that this works if I run it from my terminal! :grimacing: I can only reproduce this error via `nix build` :(

**Comment by Ericson2314** [comment]
Try running that command without network access? I bet it's trying to download stuff to make a lock file on the fly.

**Comment by FintanH** [comment]
> Try running that command without network access? I bet it's trying to download stuff to make a lock file on the fly.

Hmmmm you're right! Why is that an issue?

**Comment by FintanH** [comment]
We _do_ have a set of binaries that seem open nix flaking since we have a `Cargo.lock` there, but unfortunately I ran into this number https://github.com/NixOS/nix/pull/4638 :upside_down_face:

**Comment by Ericson2314** [comment]
Yes the fact of the matter is you want to keep 1 or more Cargo.locks that don't go to `crates.io` when the library is published, but are used for CI. I general. packaged have "extract stuff" that isn't published for downstream consumption, but useful for development.

I dislike flakes, but anyways flake problems are anyways out of scope here. https://github.com/NixOS/nix/pull/4638 you can avoid but just writing down the reference? But `niv` and `nix-thunk` shouldn't suffer from it.
